### PR TITLE
Async await code refactoring

### LINKED ISSUE #884: Async-await refactoring
Research common async-await antipatterns and mistakes, find them in project and fix it.

We made one PR(#889) but still looks like more research is needed.
For example we should check asynchronous calls in loops and optimize them, like in PR #889.

### PR DESCRIPTION
**Pull request by PavelTsyposhchuk**
Refactoring async-awaits antipatterns and mistakes

### CODE REVIEW

File: CharlieBackend.Business/Services/StudentGroupService.cs [hunk 1]
```diff
@@ -241,10 +241,11 @@ public async Task<Result<IList<StudentGroupDto>>> GetAllStudentGroupsAsync(DateT
 
         public async Task<bool> DeleteStudentGroupAsync(long StudentGroupId)
         {
-            var lessonsOfStudentGroup = _unitOfWork.LessonRepository.GetAllLessonsForStudentGroup(StudentGroupId);
+            var lessonsOfStudentGroup = await _unitOfWork.LessonRepository.GetAllLessonsForStudentGroup(StudentGroupId);
```
Comment (DevPew01) [inline CharlieBackend.Business/Services/StudentGroupService.cs]: Maybe need Async here?

File: CharlieBackend.Panel/Controllers/HomeworksController.cs [hunk 2]
```diff
@@ -60,29 +60,36 @@ public async Task<IActionResult> SelectGroups(long courseId)
...
+                    listTasks.Add(_mentorService.GetMentorByIdAsync(id));
                 }
+
+                IList<MentorEditViewModel> mentors = await Task.WhenAll(listTasks);
+
                 return View("Step4Mentors", mentors);
             }
+
             IEnumerable<LessonViewModel> lessons = allLessons.Where(x => x.StudentGroupId == stGroupId)
                 .Where(z => z.ThemeName == themeN).Where(d => d.MentorId == mentorId).ToList();
```
Comment (DevPew01) [inline CharlieBackend.Panel/Controllers/HomeworksController.cs]: In my opinion, this multiple delegates call  Where(...).Where(...)... in linq-to-objects more slower then one condition with &&  operators.

**Review by BobTheSuperman — APPROVED** [review]

**Review by Ssashaa — APPROVED** [review]

**Review by BohdanZaliubovskyi — APPROVED** [review]
Look's good for me.

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Csv/StudentsGroupCsvFileExport.cs [hunk 3]
```diff
@@ -29,12 +29,7 @@ private void FillFile(IEnumerable<Student> students)
 
             byte[] byteLine = line.ToString().ConvertLineToArray();
 
-            _memoryStream.Write(byteLine);
-        }
-
-        public override async Task FillFileAsync(IEnumerable<Student> data)
-        {
-            await Task.Run(() => FillFile(data));
+            await _memoryStream.WriteAsync(byteLine);
```
Comment (andy-dendy-repo) [inline CharlieBackend.Business/Services/FileServices/ExportFileServices/Csv/StudentsGroupCsvFileExport.cs]: why don't you remove async and return _memoryStream.WriteAsync(byteLine); ?

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookHtmlFileExport.cs [hunk 4]
```diff
@@ -72,11 +72,11 @@ private void FillPresences(StudentsClassbookResultDto data)
 
                 byte[] byteLine = html.ToString().ConvertLineToArray();
 
-                _memoryStream.Write(byteLine);
+                await _memoryStream.WriteAsync(byteLine);
```
Comment (andy-dendy-repo) [inline CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookHtmlFileExport.cs]: the same with async and return

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookHtmlFileExport.cs [hunk 5]
```diff
@@ -131,7 +131,7 @@ private void FillMarks(StudentsClassbookResultDto data)
...
```
Comment (andy-dendy-repo) [inline CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/ClassbookHtmlFileExport.cs]: the same

**Review by DevPew01 — APPROVED** [review]

**Review by fllimon — CHANGES_REQUESTED** [review]
Looks good

**Review by fllimon — APPROVED** [review]

**Pull request merged**